### PR TITLE
Add a list of BWC versions for CI

### PR DESCRIPTION
**Pull request by alpar-t**
This PR adds a list of index compatible versions to the `.ci` directory
as well as a way to generate and verify it.

Unfortunately there is no easy way in Jenkins to have the build generate then
consume this YML axis config.
I hate that this would need maintenance on versions bumps, but the
potential benefit here is reducing the bwc builds that can take more than
24 hours to less than 20 minutes.
This is possible because the CI setup would use a matrix job to run
something like:
```
./graldew 'v7.0.0#bwcTest'
```
For every index compatible version.
On top of that `--parallel` should be possible even without testclusters
due to the limited number of clusters being set up here.

The example command above runs in exactly 10 minutes on my laptop,
thus I'm proposing to accept this compromise while we work out the
infra to do this more dynamically.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -107,6 +107,17 @@ subprojects {
  * logic in VersionUtils.java. */
 BwcVersions versions = new BwcVersions(file('server/src/main/java/org/elasticsearch/Version.java').readLines('UTF-8'))
 
+task updateCIBwcVersions() {
+    doLast {
+        File yml = file(".ci/bwcVersions")
+        yml.text = ""
+        yml << "BWC_VERSION:\n"
```
Comment (mark-vieira) [inline build.gradle]: You could simplify this by just dong `yml.text = "BWC_VERSION:\n"`

**Review by mark-vieira — APPROVED** [review]
LGTM. Keeping the list of BWC versions is not ideal but if it lets us break up the BWC build then I think it's a net-win. We can potentially look at ways to lower the maintenance overhead here.

**Comment by rjernst** [comment]
I wonder if we need every concrete version, or if instead we could add minor version tasks to run the concrete tasks. So eg we could have `v7.2#bwcTest` to test all the versions fro 7.2. In practice we usually only have a couple versions for a release branch. This wouldn't allow the same parallelism, but it would simplify when this file needs to be modified (only on branch cutting instead of on all version bumps).

**Comment by mark-vieira** [comment]
I think that's a good compromise and that way we don't have to touch this after every patch release.

**Comment by atorok** [comment]
> I wonder if we need every concrete version, or if instead we could add minor version tasks to run the concrete tasks. So eg we could have `v7.2#bwcTest` to test all the versions fro 7.2. In practice we usually only have a couple versions for a release branch. This wouldn't allow the same parallelism, but it would simplify when this file needs to be modified (only on branch cutting instead of on all version bumps).

I don't really like the disproportion this would create for bugfix releases on branches that need to be backwards compatible with them, especially that those jobs will have to run wire compatibility tests in addition to index comparability.
I would rather look at fixing the maintenance burden here. Possibly by (ab)using the [EnvInject Plugin](https://wiki.jenkins.io/display/JENKINS/EnvInject+Plugin)  that can run a script after checkout. Or by installing and using the Groovy extension for matrix projects. 

Let's check this in and see how we like it first, we can make a final call before we pull the plug on the old bwc jobs. Initially we\ll just add new jobs.

**Comment by atorok** [comment]
To exemplify the comment above, we have a `5.6.16` that we still test against in `6.8` that would run tests against 17 versions in one job.

**Comment by rjernst** [comment]
> we have a 5.6.16 that we still test against in 6.8 that would run tests against 17 versions in one job.

I think that is an extreme case we can handle separately. For example, nothing would preclude us from use per bugfix release jobs when we get to the maintenance branch.

**Comment by mark-vieira** [comment]
There's also not a huge benefit to these builds running faster, compared to other builds like intake. This is increasingly less so for older branches, as they are less frequently worked on. That is, naturally, newer branches will have fewer patch releases. So by extension, the more patch releases a minor has, the less likely for us to be touching it. We also release minor versions pretty frequently so I agree with Ryan that the 5.6.x case is a bit pathological.

Essentially, these are periodic builds so their runtime doesn't affect developer productivity so much. If the BWC builds for older release branches continue to be slow I think that's fine if the trade-off is less overall build maintenance.

**Pull request merged**